File: backend/schemas/drift.py
"""
Pydantic schemas for the Drift Detection domain.

Covers:
  - Drift settings (request / response)
  - Drift check response
  - Drift summary response
  - Trigger drift check request
"""

from datetime import datetime

from pydantic import BaseModel

# =============================================================================
# Drift Settings Schemas
# =============================================================================

class DriftSettingsRequest(BaseModel):
    """Request model for drift settings."""
    enabled: bool = False
    schedule_type: str = "cron"
    schedule_value: str = "0 2 * * *"
    check_all_modules: bool = True
    module_ids: list[int] | None = None
    notify_on_drift: bool = True
    notification_channels: list[str] | None = None
    notification_config: dict | None = None
    ignore_insignificant_changes: bool = False
    ignore_patterns: list[str] | None = None


class DriftSettingsResponse(BaseModel):
    """Response model for drift settings."""
    id: int
    project_id: int
    enabled: bool
    schedule_type: str
    schedule_value: str
    check_all_modules: bool
    module_ids: list[int] | None
    notify_on_drift: bool
    notification_channels: list[str] | None
    notification_config: dict | None
    ignore_insignificant_changes: bool
    ignore_patterns: list[str] | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# Drift Check Schemas
# =============================================================================

class DriftCheckResponse(BaseModel):
    """Response model for drift check."""
    id: int
    project_id: int
    module_id: int | None
    module_name: str | None
    schedule_enabled: bool
    schedule_cron: str | None
    last_check_at: datetime | None
    next_check_at: datetime | None
    drift_detected: bool
    drift_summary: str | None
    drift_details: dict | None
    task_id: int | None
    status: str
    error_message: str | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# =============================================================================
# Drift Summary / Trigger Schemas
# =============================================================================

class DriftSummaryResponse(BaseModel):
    """Response model for drift summary."""
    total_checks: int
    drift_detected_count: int
    no_drift_count: int
    failed_count: int
    last_check_at: datetime | None
    projects_with_drift: int
    modules_with_drift: int


class TriggerDriftCheckRequest(BaseModel):
    """Request model for triggering drift check."""
    module_ids: list[int] | None = None

File: frontend-v2/src/types/drift.ts
// Drift Detection types

import type { JsonValue } from './common';
import type { DeploymentEngineType } from './modules';

export interface DriftSettings {
  id: number;
  project_id: number;
  enabled: boolean;
  schedule_type: "cron" | "interval";
  schedule_value: string;
  check_all_modules: boolean;
  module_ids?: number[];
  notify_on_drift: boolean;
  notification_channels?: string[];
  notification_config?: Record<string, JsonValue>;
  ignore_insignificant_changes: boolean;
  ignore_patterns?: string[];
  created_at: string;
  updated_at: string;
}

export interface DriftSettingsRequest {
  enabled?: boolean;
  schedule_type?: "cron" | "interval";
  schedule_value?: string;
  check_all_modules?: boolean;
  module_ids?: number[];
  notify_on_drift?: boolean;
  notification_channels?: string[];
  notification_config?: Record<string, JsonValue>;
  ignore_insignificant_changes?: boolean;
  ignore_patterns?: string[];
}

export interface DriftCheck {
  id: number;
  project_id: number;
  module_id?: number;
  module_name?: string;
  schedule_enabled: boolean;
  schedule_cron?: string;
  last_check_at?: string;
  next_check_at?: string;
  drift_detected: boolean;
  drift_summary?: string;
  drift_details?: {
    drift_detected: boolean;
    resource_changes: {
      add: number;
      change: number;
      destroy: number;
    };
    changed_resources: Array<{
      address: string;
      action: string;
    }>;
    summary: string;
  };
  task_id?: number;
  status: "scheduled" | "checking" | "completed" | "failed";
  error_message?: string;
  created_at: string;
  updated_at: string;
}

export interface DriftSummary {
  total_checks: number;
  drift_detected_count: number;
  no_drift_count: number;
  failed_count: number;
  last_check_at?: string;
  projects_with_drift: number;
  modules_with_drift: number;
}

export interface TriggerDriftCheckRequest {
  module_ids?: number[];
}

export interface RecentDriftedItem {
  id: number;
  project_id: number;
  project_name: string;
  module_id: number;
  module_name: string;
  drift_summary: string | null;
  drift_details: DriftCheck['drift_details'] | null;
  last_check_at: string | null;
  resource_changes: {
    add: number;
    change: number;
    destroy: number;
  } | null;
  changed_resources: Array<{
    address: string;
    action: string;
  }>;
}

export interface ClusterDriftModuleStatus {
  module_id: number;
  module_name: string;
  module_path: string;
  engine_type: DeploymentEngineType;
  status: 'ok' | 'drift' | 'unchecked';
  drift_detected: boolean;
  drift_summary: string | null;
  drift_details: Record<string, unknown> | null;
  last_check_at: string | null;
  check_id: number | null;
}

export interface ClusterDriftStatus {
  cluster_id: number;
  project_id: number;
  drift_enabled: boolean;
  total_modules: number;
  modules_with_drift: number;
  modules_ok: number;
  modules_unchecked: number;
  overall_status: 'ok' | 'drift_detected' | 'unchecked' | 'partial';
  module_statuses: ClusterDriftModuleStatus[];
}
